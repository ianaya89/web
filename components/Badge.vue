<template>
  <div
    class="rounded-full bg-red text-white w-6 h-6 text-center text-p4 flex items-center border-white border-solid border-2"
    :class="{'change-value-animation':valueChanged}"
  >
    <transition name="fade" mode="out-in">
      <span v-if="value !== null" :key="badgeValue" class="pt-px flex-1 font-bold leading-none">
        {{ badgeValue }}
      </span>
    </transition>
  </div>
</template>

<script>
export default {
  name: 'Badge',

  props: {
    value: {
      type: Number,
      default: null
    }
  },

  data: () => ({
    valueChanged: false
  }),

  computed: {
    badgeValue() {
      return this.value > 99 ? '99+' : this.value
    }
  },

  watch: {
    value() {
      this.valueChanged = true
      setTimeout(() => {
        this.valueChanged = false
      }, 1000)
    }
  }
}
</script>

<style lang="postcss" scoped>
@keyframes badge-bounce {
  0% {
    transform: scale(1, 1) translate3d(0, 0, 0);
  }
  10% {
    transform: scale(1.15, 0.85) translate3d(0, 0, 0);
  }
  30% {
    transform: scale(0.9, 1.1) translate3d(0, -40%, 0);
  }
  50% {
    transform: scale(1.05, 0.95) translate3d(0, 0, 0);
  }
  60% {
    transform: scale(0.95, 1.05) translate3d(0, -15%, 0);
  }
  70% {
    transform: scale(1.02, 0.98) translate3d(0, 0, 0);
  }
  75% {
    transform: scale(1, 1) translate3d(0, 0, 0);
  }
  100% {
    transform: scale(1, 1) translate3d(0, 0, 0);
  }
}

.change-value-animation {
  animation-duration: 1000ms;
  animation-iteration-count: 1;
  animation-name: badge-bounce;
  animation-timing-function: ease;
  transform-origin: bottom;
}
</style>
