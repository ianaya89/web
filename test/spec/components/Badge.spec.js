import flushPromises from 'flush-promises'
import { shallowMount } from '@vue/test-utils'
import Badge from '@/components/Badge'

describe('Badge', () => {
  it('renders empty badge', () => {
    const wrapper = shallowMount(Badge)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders badge value 3', () => {
    const wrapper = shallowMount(Badge, {
      propsData: {
        value: '3'
      }
    })

    const badgeDiv = wrapper.find('div')
    expect(badgeDiv.find('span').exists()).toBe(true)
    expect(badgeDiv.find('span').text()).toBe('3')
    expect(wrapper.props().value).toBe('3')
  })

  it('renders big badge value', () => {
    const wrapper = shallowMount(Badge, {
      propsData: {
        value: '30'
      }
    })
    expect(wrapper.element).toMatchSnapshot()
    wrapper.setProps({ value: 100 })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('check change badge value', async () => {
    const wrapper = shallowMount(Badge, {
      propsData: {
        value: '30'
      }
    })
    expect(wrapper.element).toMatchSnapshot()
    wrapper.setProps({ value: 40 })
    wrapper.vm.valueChanged = true
    await flushPromises()
    wrapper.vm.valueChanged = false
  })
})
