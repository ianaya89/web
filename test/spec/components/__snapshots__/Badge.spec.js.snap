// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge check change badge value 1`] = `
<div
  class="rounded-full bg-red text-white w-6 h-6 text-center text-p4 flex items-center border-white border-solid border-2"
>
  <transition-stub
    mode="out-in"
    name="fade"
  >
    <span
      class="pt-px flex-1 font-bold leading-none"
    >
      
      30
    
    </span>
  </transition-stub>
</div>
`;

exports[`Badge renders big badge value 1`] = `
<div
  class="rounded-full bg-red text-white w-6 h-6 text-center text-p4 flex items-center border-white border-solid border-2"
>
  <transition-stub
    mode="out-in"
    name="fade"
  >
    <span
      class="pt-px flex-1 font-bold leading-none"
    >
      
      30
    
    </span>
  </transition-stub>
</div>
`;

exports[`Badge renders big badge value 2`] = `
<div
  class="rounded-full bg-red text-white w-6 h-6 text-center text-p4 flex items-center border-white border-solid border-2"
>
  <transition-stub
    mode="out-in"
    name="fade"
  >
    <span
      class="pt-px flex-1 font-bold leading-none"
    >
      
      30
    
    </span>
  </transition-stub>
</div>
`;

exports[`Badge renders empty badge 1`] = `
<div
  class="rounded-full bg-red text-white w-6 h-6 text-center text-p4 flex items-center border-white border-solid border-2"
>
  <transition-stub
    mode="out-in"
    name="fade"
  >
    <!---->
  </transition-stub>
</div>
`;
